web/src/ui.rs

```diff
@@ -1,10 +1,99 @@
 use super::JavascriptPlayer;
-use ruffle_core::backend::ui::{MouseCursor, UiBackend};
+use rfd::{AsyncFileDialog, FileHandle};
+use ruffle_core::backend::ui::{
+    DialogResultFuture, Error, FileDialogResult, FileFilter, MouseCursor, UiBackend,
+};
 use ruffle_core::events::KeyCode;
 use ruffle_web_common::JsResult;
 use std::collections::HashSet;
+use std::path::Path;
 use web_sys::{HtmlCanvasElement, KeyboardEvent};
 
+use chrono::{DateTime, Utc};
+
+#[cfg(target_arch = "wasm32")]
+use chrono::NaiveDateTime;
+
+pub struct WebFileDialogResult {
+    handle: Option<FileHandle>,
+}
+
+impl WebFileDialogResult {
+    pub fn new(handle: Option<FileHandle>) -> Self {
+        Self { handle }
+    }
+}
+
+fn get_extension_from_filename(filename: &str) -> Option<String> {
+    Path::new(filename)
+        .extension()
+        .and_then(|x| x.to_str())
+        .map(|x| ".".to_owned() + x)
+}
+
+impl FileDialogResult for WebFileDialogResult {
+    fn is_cancelled(&self) -> bool {
+        self.handle.is_none()
+    }
+
+    // For some reason the test suite compiles this code.
+    #[cfg(not(target_arch = "wasm32"))]
+    fn creation_time(&self) -> Option<DateTime<Utc>> {
+        unreachable!();
+    }
```

Comment:
Let's make two separate `impl` blocks and `#[cfg(not(target_arch = "wasm32"))]` on the impl block, instead of doing `#[cfg]` on each method.

Reply:
There are functions that should be in both configs and functions that should be in just one. Since I can't do two trait implemenrations, it means I have to duplicate the functions that are in both... advice?